src/main/java/org/benetech/servicenet/service/impl/ActivityFilterServiceImpl.java

```diff
@@ -67,8 +70,15 @@ public ActivityFilterServiceImpl(GeocodingResultRepository geocodingResultReposi
     }
 
     @Override
-    public Set<String> getTaxonomies() {
-        return taxonomyRepository.findAll().stream().map(Taxonomy::getName).collect(Collectors.toSet());
+    public Map<String, Set<String>> getTaxonomies() {
+        List<Taxonomy> taxonomies = taxonomyRepository.findAssociatedTaxonomies();
+        Map<String, Set<String>> taxonomiesByProvider = new HashMap<>();
+        for (Taxonomy taxonomy : taxonomies) {
+            Set<String> providersTaxonomies = taxonomiesByProvider.getOrDefault(taxonomy.getProviderName(), new HashSet<>());
+            providersTaxonomies.add(taxonomy.getName());
+            taxonomiesByProvider.put(taxonomy.getProviderName(), providersTaxonomies);
```

Comment:
If performance is a concern here, this can changed to just adding new elements to the Value of the Map. If not - it is fine.